src/internal/i18n/__tests__/i18n.test.tsx

```diff
@@ -0,0 +1,173 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import React from 'react';
+import { render } from '@testing-library/react';
+import { useInternalI18n } from '../context';
+import { I18nProvider, I18nProviderProps } from '../provider';
+
+interface TestComponentProps {
+  topLevelString?: string;
+  topLevelFunction?: (type: 'function') => string;
+  nested?: {
+    nestedString?: string;
+    nestedFunction?: (props: { type: 'function' }) => string;
+  };
+}
+
+const MESSAGES: I18nProviderProps.Messages = {
+  '@cloudscape-design/components': {
+    en: {
+      'test-component': {
+        topLevelString: 'top level string',
+        topLevelFunction: 'top level {type}',
+        'nested.nestedString': 'nested string',
+        'nested.nestedFunction': 'nested {type}',
+      },
+    },
+  },
+};
+
+function TestComponent(props: TestComponentProps) {
+  const format = useInternalI18n('test-component');
+
+  return (
+    <ul>
+      <li id="top-level-string">{format('topLevelString', props.topLevelString)}</li>
+      <li id="top-level-function">
+        {format('topLevelFunction', props.topLevelFunction, format => type => format({ type }))?.('function')}
+      </li>
+
+      <li id="nested-string">{format('nested.nestedString', props.nested?.nestedString)}</li>
+      <li id="nested-function">
+        {format(
+          'nested.nestedFunction',
+          props.nested?.nestedFunction,
+          format => props => format(props)
+        )?.({ type: 'function' })}
+      </li>
+    </ul>
+  );
+}
+
+it('detects locale from the html tag', () => {
+  document.documentElement.lang = 'es';
```

Comment:
I think this would affect other test cases, as well. Probably needs to be reset.